cluster-autoscaler/core/scaleup/orchestrator/executor.go

```diff
@@ -0,0 +1,240 @@
+	gpuResourceName, gpuType := gpu.GetGpuInfoForMetrics(gpuConfig, availableGPUTypes, nodeInfo.Node(), nil)
+	klog.V(0).Infof("Scale-up: setting group %s size to %d", info.Group.Id(), info.NewSize)
+	e.autoscalingContext.LogRecorder.Eventf(apiv1.EventTypeNormal, "ScaledUpGroup",
+		"Scale-up: setting group %s size to %d instead of %d (max: %d)", info.Group.Id(), info.NewSize, info.CurrentSize, info.MaxSize)
+	increase := info.NewSize - info.CurrentSize
+	if err := info.Group.IncreaseSize(increase); err != nil {
+		e.autoscalingContext.LogRecorder.Eventf(apiv1.EventTypeWarning, "FailedToScaleUpGroup", "Scale-up failed for group %s: %v", info.Group.Id(), err)
+		aerr := errors.ToAutoscalerError(errors.CloudProviderError, err).AddPrefix("failed to increase node group size: ")
+		e.clusterStateRegistry.RegisterFailedScaleUp(info.Group, metrics.FailedScaleUpReason(string(aerr.Type())), now)
+		return aerr
+	}
+	e.clusterStateRegistry.RegisterOrUpdateScaleUp(
+		info.Group,
+		increase,
+		time.Now())
+	metrics.RegisterScaleUp(increase, gpuResourceName, gpuType)
+	e.autoscalingContext.LogRecorder.Eventf(apiv1.EventTypeNormal, "ScaledUpGroup",
+		"Scale-up: group %s size set to %d instead of %d (max: %d)", info.Group.Id(), info.NewSize, info.CurrentSize, info.MaxSize)
+	return nil
+}
+
+func combineConcurrentScaleUpErrors(errs []errors.AutoscalerError) errors.AutoscalerError {
+	if len(errs) == 0 {
+		return nil
+	}
+	if len(errs) == 1 {
+		return errs[0]
+	}
+	// sort to stabilize the results and easier log aggregation
+	sort.Slice(errs, func(i, j int) bool {
+		errA := errs[i]
+		errB := errs[j]
+		if errA.Type() == errB.Type() {
+			return errs[i].Error() < errs[j].Error()
+		}
+		return errA.Type() < errB.Type()
+	})
+	firstErrMsg := errs[0].Error()
+	firstErrType := errs[0].Type()
+	singleError := true
+	singleType := true
+	for _, err := range errs {
+		singleType = singleType && firstErrType == err.Type()
+		singleError = singleError && singleType && firstErrMsg == err.Error()
+	}
+	if singleError {
+		return errs[0]
+	}
+	var builder strings.Builder
+	builder.WriteString(firstErrMsg)
+	builder.WriteString(" ...other concurrent errors: [")
+	concurrentErrs := make(map[string]bool)
+	for _, err := range errs {
+		var concurrentErr string
+		if singleType {
+			concurrentErr = err.Error()
+		} else {
+			concurrentErr = fmt.Sprintf("[%s] %s", err.Type(), err.Error())
+		}
+		n := len(concurrentErrs)
+		if !concurrentErrs[concurrentErr] && n > 0 {
+			if n > 1 {
+				builder.WriteString(", ")
+			}
+			builder.WriteString(fmt.Sprintf("%q", concurrentErr))
+		}
+		concurrentErrs[concurrentErr] = true
+	}
+	builder.WriteString("]")
+	return errors.NewAutoscalerError(firstErrType, builder.String())
+}
+
+// Asserts all groups are scaled only once.
+// Scaling one group multiple times concurrently may cause problems.
+func assertUniqueNodeGroups(scaleUpInfos []nodegroupset.ScaleUpInfo) errors.AutoscalerError {
```

Comment:
I think we usually use `assert` in context of unit tests, I would use `check` or something similar in this context.